logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/search/SearchUsersUseCase.kt

```diff
@@ -0,0 +1,104 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+package com.wire.kalium.logic.feature.search
+
+import com.wire.kalium.logic.data.id.ConversationId
+import com.wire.kalium.logic.data.publicuser.SearchUserRepository
+import com.wire.kalium.logic.data.publicuser.model.UserSearchDetails
+import com.wire.kalium.logic.data.user.UserId
+import com.wire.kalium.logic.functional.getOrElse
+import com.wire.kalium.logic.functional.map
+
+/**
+ * Use case for searching users.
+ */
+class SearchUsersUseCase internal constructor(
+    private val searchUserRepository: SearchUserRepository,
+    private val selfUserId: UserId
+) {
+    suspend operator fun invoke(
+        searchQuery: String,
+        excludingConversation: ConversationId?,
+        customDomain: String?
+    ): Result {
+        return if (searchQuery.isBlank()) {
+            Result(
+                connected = searchUserRepository.initialSearchList(excludingConversation).getOrElse(emptyList()),
+                notConnected = emptyList()
+            )
+        } else {
+            handleSearch(searchQuery, excludingConversation, customDomain)
+        }
+    }
+
+    private suspend fun handleSearch(
+        searchQuery: String,
+        excludingConversation: ConversationId?,
+        customDomain: String?
+    ): Result {
+        val localSearchResult =
+            searchUserRepository.searchLocalByName(searchQuery, excludingConversation)
+                .getOrElse(emptyList())
+                .associateBy { it.id }
+                .toMutableMap()
+
+        val remoteResults =
+            searchUserRepository.searchUserDirectory(searchQuery, customDomain ?: selfUserId.domain, MAX_SEARCH_RESULTS)
+                .map { userSearchResult ->
+                    userSearchResult.result.map {
+                        UserSearchDetails(
+                            id = it.id,
+                            name = it.name,
+                            completeAssetId = it.completePicture,
+                            previewAssetId = it.previewPicture,
+                            type = it.userType,
+                            connectionStatus = it.connectionStatus
+                        )
+                    }
+                }.getOrElse(
+                    emptyList()
+                ).associateBy { it.id }
+                .toMutableMap()
```

Comment:
This can be parallelised by wrapping it in a `coroutineScope { }` and using `async {}`

Then merge local + remote to proceed.